File: sumtoRoot.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
"""
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no
"""
class Solution:
    def sumNumbers(self, root: TreeNode) -> int:
        
        
        def helper(s, root):
            
            if root is None:
                return 0

            s = 10 * s +  root.val
            
            if not root.left and not root.right:
                return s
            
            return helper(s, root.left) + helper(s, root.right)
            
                                        
        return helper(0, root)

File: sumToRoot.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : None.


// Your code here along with comments explaining your approach
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
	public int sumNumbers(TreeNode root) {
		 return helper(root, 0);
	}
	
	private int helper(TreeNode root, int sum) {
		 if (root == null) return 0;
		 
		 if (root.left == null && root.right == null) {
			  sum = sum*10 + root.val;
			  return sum;
		 }
		 
		 return helper(root.left, sum*10+root.val) + helper(root.right, sum*10+root.val);
	}
}